thetis/adr_model.py

```diff
@@ -0,0 +1,220 @@
+import firedrake as fd
+import networkx as nx
+import sympy
+from sympy.parsing.sympy_parser import \
+    convert_xor, \
+    parse_expr, \
+    standard_transformations
+from sympy.utilities.lambdify import lambdify
+
+
+class ADR_Model:
+    """
+    Data structure to store an ADR model specification.
+
+    Performs validation checks on a user-provided dictionary of strings
+    and parses strings representing mathematical expressions into
+    SymPy expressions.
+
+    :cvar transformations: Transformations applied to strings parsed as
+        SymPy expressions.
+    :type transformations: tuple
+    :ivar constants:  Dictionary mapping strings used to uniquely identify
+        constants to the values of those constants.
+    :type constants: dict
+    :ivar species: Dictionary mapping strings used to uniquely identify
+        biological or chemical species to dictionaries that contain data
+        pertaining to those species.
+    :type species: dict
+    """
+
+    # convert_xor causes "x^2" to be parsed as "x**2"
+    transformations = (
+        standard_transformations + (convert_xor,))
+
+    reserved_symbols = {
+        "pi", "e",
+        "x", "y", "z", "t",
+        "u", "u_x", "u_y", "u_z"}
```

Comment:
These symbols are reserved to allow reaction terms to contain pi, e, spatial coordinates, time, the flow field, and individual components of the flow field.

I'm not aware of any issues with pi or e, because these are just constants. However, at present, `parse_tracers_from_dict` will not work properly if it encounters a reaction term containing any of these other symbols. It expects all symbols to correspond to either constants or species.

The simplest way to solve this problem would be to remove everything except pi and e from `reserved_symbols`, but it might be worth considering alternatives it if people would like to be able to include spatial coordinates, etc. in reaction terms.